pages/chaoss/visualizations/contrib_importance_over_time.py

```diff
@@ -0,0 +1,423 @@
+    # subset df such that the rows correspond to the window of time defined by period from and period to
+    mask = (df['created_at'] >= period_from) & (df['created_at'] <= period_to)
+    df_in_range = df.loc[mask]
+    
+    # initialize varibles to store contributor importance accoding to action type
+    commit, issueOpened, prOpened, prReview, prComment = None, None, None, None, None
+
+    # count the number of contributions each contributor has made according each action type
+    df_count_cntrbs = df_in_range.groupby(['Action', 'cntrb_id'])['cntrb_id'].count().to_frame()
+    df_count_cntrbs = df_count_cntrbs.rename(columns={'cntrb_id': 'count'}).reset_index()
+     
+    # pivot df such that the column names correspond to the different action types and the values are the number of contributions of each contributor
+    df_count_cntrbs = df_count_cntrbs.pivot(columns='Action', values='count')
+
+    commit = calc_cntrb_importance(df_count_cntrbs, 'Commit', threshold)
+    issueOpened = calc_cntrb_importance(df_count_cntrbs, 'Issue Opened', threshold)
+    prOpened = calc_cntrb_importance(df_count_cntrbs, 'PR Opened', threshold)
+    prReview = calc_cntrb_importance(df_count_cntrbs, 'PR Review', threshold)
+    prComment = calc_cntrb_importance(df_count_cntrbs, 'PR Comment', threshold)
+
+    return commit, issueOpened, prOpened, prReview, prComment
+
+def process_data(df, patterns, threshold, window_width, step_size, start_date, end_date):
+    # convert to datetime objects rather than strings
+    df["created_at"] = pd.to_datetime(df["created_at"], utc=True)
+    
+    # order values chronologically by created_at date
+    df = df.sort_values(by='created_at', ascending=True)
+
+    # if the start_date and/or the end date is not specified set them to the beginning and most recent created_at date
+    if start_date is None:
+        start_date = df["created_at"].min()
+    if end_date is None:
+        end_date = df["created_at"].max()
+
+    if patterns:
+        # remove rows where Login column value contains the substring 'bot'
+        mask = df['login'].str.contains("|".join(patterns), na=False)
+        df = df[~mask]
+    
+    # threshold is an integer value eg. 10, 20,..., 90 since dcc.Slider only accepts integers as values
+    # divide by 10 to convert it to a decimal representation of a percentage eg. 0.10, 0.20,..., 0.90
+    threshold = threshold / 10
+    
+    # create bins with a size equivalent to the the step size starting from the start date up to the end date
+    period_from = pd.date_range(start=start_date, end=end_date, freq=f'{step_size}m', inclusive="both")
+    # store the period_from dates in a df 
+    df_final = period_from.to_frame(index=False, name="period_from")
+    # calculate the end of each interval and store the values in a column named period_from
+    df_final['period_to'] = df_final.apply(lambda row: row.period_from + relativedelta(months = window_width), axis=1)
```

Comment:
use pd.date_offset instead

https://stackoverflow.com/questions/46741423/add-months-to-a-date-in-pandas
https://pandas.pydata.org/docs/reference/api/pandas.tseries.offsets.DateOffset.html